FEA Add Oblique trees and oblique splitters to tree module: Enables extensibility of the trees

adam2392 opened the pull request:
#### Reference Issues/PRs
Closes: #20819 

This is ready for review.

#### What does this implement/fix? Explain your changes.

Implements oblique trees/splitter as a subclass of the existing `Tree`/`Splitter`. This further allows extensibility of the sklearn `Tree` and `Splitter` to allow downstream packages that have to define more complex and exotic sampling mechanisms. The total Cython code that constitutes the logic addition is around ~950 LOC, while the rest are from unit tests, and adding the Python API.

This is an extension of decision trees to generalize to random linear combination of feature oblique trees proposed by Breimann in 2001. This will enable 3rd party packages to subclass this code to instantiate other trees that work on "combinations" of features in some way (e.g. taking the weighted sum, or a kernel). The general intuition is that OF can sample more diverse set of features enabling better generalization and robustness to high-dimensional noise. RF should be used if the user suspects all data is aligned on the feature dimensions. The tradeoffs of computation time, model size and scoring time are complex because OF can fit shorter trees, but requires storage of additional data. OF needs to perform additional operations increasing computation time, but it can be negligible. We always suggest using OF first if you have the computational and storage flexibility. We suggest using RF for now if you have very large sample sizes and very very strict runtime/storage constraints.

- **Inclusion criterion** was discussed. Breiman 2001, proposed oblique forests and demonstrated its superiority. Also see https://github.com/scikit-learn/scikit-learn/issues/20819#issuecomment-1065523071
- **Maintainability** was addressed by figuring out a way that `ObliqueSplitter` and `ObliqueTree` can subclass the existing code. 
- **Runtime performance** is benchmarked (https://gist.github.com/adam2392/a652aa1c88ba94f0d6aab96ccd3ade24)
- **Interpretability** is demonstrated using an extension of the RandomForest example with `iris` dataset (https://gist.github.com/adam2392/78519091104cecfeb8ff796eac7e8115)

##### Experiments supporting the change

Requested by Thomas Fan, Guillaume, and Olivier, we ran the following experiments.

- runtime: see above
- visual interpretation of the decision boundaries: see iris example
- grid search outcomes on simulated data of OF vs RF: https://nbviewer.org/github/neurodata/scikit-learn/blob/5e8f1a6a506bf77fc99cdb59d81fe02f8a301e20/notebook/simulation_benchmark_OF_vs_RF.ipynb
- grid search outcomes on low-dimension MNIST of OF vs RF: https://nbviewer.org/github/neurodata/scikit-learn/blob/25a1e9819c5def42f4f919eafdf4df4b065b2c08/notebook/mnist_benchmark_OF_vs_RF.ipynb
- performance (accuracy) vs score time: see above 2 notebooks
- performance vs fit time: see above 2 notebooks
- performance vs model size: see above 2 notebooks
- grid search outcomes on an openml dataset: TBD

##### Docs changes for education and user-guidance

- [x] extension of the `examples/tree/plot_iris_dtc.py` to include oblique trees
- [x] `modules/tree.rst` with a new section on oblique trees
- [x] `modules/ensemble.rst` with a new section on oblique forests
- [ ] A new example in `examples/ensemble/plot_oblique_axis_aligned_forests.py` demonstrating oblique forest superiority on a real dataset from openml and a simulated dataset

The changes to `setup.py` files were necessary to compile package in some of the CI pipelines. It worked for me w/o it locally on my Mac M1 machine though.

#### Tracking Progress on sub-items - used for keeping track of challenges that arose (optional read)
The remaining items to complete are:
- [x] Add unit tests
- [x] Refactor to use memory views as in normal splitter (https://github.com/scikit-learn/scikit-learn/pull/23273/files)
- [x] Fix pickling issue in https://github.com/scikit-learn/scikit-learn/pull/22754#issuecomment-1128124308
- [x] Add `feature_importances`, or specify an error message if it's not allowed.
- [x] Add extension to the RF-Iris dataset example done by Jong Shin within the docs. Also adding a synthetic dataset perhaps that looks cleaner. And another dataset, for example `digits` (8 x 8 images) that can get shallower/smaller-forests vs RandomForest; e.g. train OF/RF with 100 trees, and subsample the forest and plot accuracy and plot the max_depth.  Basically get a few more examples that are candidates to include into the docs: 1 example showing an intuitive visualization and then another example to demonstrate possible superiority in terms of real example. E.g. digits horizontally stack the real signal but then create permuted noisy copies and stack again. Put it into a separate notebook for now for downstream integration into scikit-learn. End goal: education for users on the inductive bias for oblique trees and when to use.
- [x] Determine if minor refactoring is needed with `TreeBuilder/Tree` for adding nodes (see: https://github.com/scikit-learn/scikit-learn/pull/22756#issuecomment-1065549821)

#### Any other comments?
This is the PR to demonstrate the end product. To simplify the merging process, I propose the merging of this PR in a series of other smaller PRs.

- [x] ~#22753~
- [x] Fix unit test organization for pickling trees: https://github.com/scikit-learn/scikit-learn/pull/22915#pullrequestreview-918823308
- [x] Small PR to update `Tree` abstractions, such as in `sklearn/tree/_classes.py` and `sklearn/tree/_tree.pyx`
- [x] Simplifications of the utility code, such as sorting to use c++ (https://github.com/scikit-learn/scikit-learn/pull/22760)

Another misc. note is that oblique and decision trees can be further easily improved in terms of training speed via quantization of continuous feature values in the data (i.e. same idea used in histogram gradient boosting).

adam2392 committed:
Adding makefile

adam2392 committed:
Adding changes to make refactor easy

adam2392 committed:
Adding changes to make refactor easy

adam2392 committed:
Whoot

adam2392 committed:
Removing fluff

adam2392 committed:
Removing fluff

adam2392 committed:
Almost working

adam2392 committed:
Almost mwe with only one nan

adam2392 committed:
Working version with subclass

adam2392 committed:
Remove jupyter notebook

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Removing scruff

adam2392 committed:
Merge branch 'obliquepr' of github.com:neurodata/scikit-learn into obliquepr

adam2392 committed:
Fix with thomas suggestions

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Merged main. Working version

adam2392 committed:
Co-author commit. Adding co-authors who helped out with code, examples, etc.

adam2392 commented:
Linking convo: https://github.com/neurodata/scikit-learn/pull/16

adam2392 committed:
Co-authored-by: Thomas Fan <thomasjpfan@gmail.com>
Co-authored-by: Chester Huynh <chester.huynh924@gmail.com>
Co-authored-by: Parth Vora <pvora4@jhu.edu>

adam2392 committed:
Try to fix setup.

Co-Authored-By: Thomas Fan <thomasjpfan@gmail.com>
Co-Authored-By: Chester Huynh <chester.huynh924@gmail.com>
Co-Authored-By: Parth Vora <pvora4@jhu.edu>

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Adding pickling

adam2392 commented:
Added Oblique trees/forests to the existing unit tests parametrizations and they mostly work, with the following issues:

- sparse dataset support, which I could possibly split into a follow-up PR?
- pickling does not work on roundtrip for some reason
- `ensemble.test_decision_path`: Not sure if this is a bug, or if oblique trees might need to be excluded from this test case

Also adding a few new tests of performance relative to `Forest_RI` (i.e. `DecisionTreeClf` and `RandomForestClf`).

adam2392 committed:
Adding working version of oblique tree/forests, along with unit
tests.

The issues still are:

- sparse dataset support, which I could possibly split into a follow-up PR?
- pickling does not work on roundtrip for some reason
- certain issues with max_leaf

adam2392 committed:
Adding almost working unit tests... Weird error with pickling

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Working version of oblique trees for all tests

adam2392 committed:
Fixed all unit tests, or at least they should work now.

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Fix lint, forest unit tests and adding user error for sparse datasets

adam2392 committed:
Fix lint, forest unit tests and adding user error for sparse datasets

adam2392 committed:
Merge branch 'obliquepr' of github.com:neurodata/scikit-learn into obliquepr

adam2392 committed:
Adding updates to fix ci

adam2392 committed:
Fix black lint

adam2392 committed:
Fix tests

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Fix import

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 commented on sklearn/tree/_splitter.pxd:
```diff
@@ -91,3 +91,7 @@ cdef class Splitter:
+
+    cdef int pointer_size(self) nogil
+
+cdef inline void sort(DTYPE_t* Xf, SIZE_t* samples, SIZE_t n) nogil
```
```suggestion
```

adam2392 commented on sklearn/tree/_utils.pxd:
```diff
@@ -44,6 +44,7 @@ ctypedef fused realloc_ptr:
     (Cell*)
     (Node**)
 
+
```
```suggestion
```

adam2392 commented on sklearn/tree/_tree.pyx:
```diff
@@ -1785,4 +1866,4 @@ cdef _build_pruned_tree(
             tree.max_depth = max_depth_seen
     if rc == -1:
-        raise MemoryError("pruning tree")
+        raise MemoryError("pruning tree")
```
```suggestion
        raise MemoryError("pruning tree")

```

adam2392 commented on sklearn/tree/_tree.pyx:
```diff
@@ -900,13 +958,18 @@ cdef class Tree:
                     if feature_to_sample[node.feature] == i:
                         feature_value = X_sample[node.feature]
-
+                        
+                        # compute the feature value to compare against threshold
+                        # X_vector = X_ndarray[i, :]
+                        # feature = self._compute_feature(X_vector, node, node_id)
```
```suggestion
```

adam2392 commented on sklearn/tree/_tree.pyx:
```diff
@@ -900,13 +958,18 @@ cdef class Tree:
                         feature_value = 0.
 
                     if feature_value <= node.threshold:
+                        node_id = node.left_child
```
```suggestion
```

adam2392 commented on sklearn/tree/_tree.pyx:
```diff
@@ -900,13 +958,18 @@ cdef class Tree:
+                        node_id = node.left_child
                         node = &self.nodes[node.left_child]
                     else:
+                        node_id = node.right_child
```
```suggestion
```

adam2392 commented on sklearn/tree/_tree.pyx:
```diff
@@ -890,6 +947,7 @@ cdef class Tree:
 
             for i in range(n_samples):
                 node = self.nodes
+                node_id = 0
```
```suggestion
```

adam2392 commented on sklearn/tree/_tree.pyx:
```diff
@@ -882,6 +936,9 @@ cdef class Tree:
         # which features are nonzero in the present sample.
         cdef SIZE_t* feature_to_sample = NULL
 
+        # to keep track of the current ID of each node
+        cdef SIZE_t node_id = 0
+
```
```suggestion
```

adam2392 commented on sklearn/tree/_oblique_tree.pxd:
```diff
@@ -0,0 +1,34 @@
+    cdef DTYPE_t _compute_feature(self, const DTYPE_t[:] X_ndarray, Node *node, SIZE_t node_id) nogil
+    cdef int _resize_c(self, SIZE_t capacity=*) nogil except -1
+
+    cpdef np.ndarray get_projection_matrix(self)
```
```suggestion
    cpdef np.ndarray get_projection_matrix(self)

```

adam2392 committed:
Apply suggestions from code review

adam2392 committed:
Removing unnecessary changesg

adam2392 committed:
Merge branch 'obliquepr' of github.com:neurodata/scikit-learn into obliquepr

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Merge branch 'obliquepr' of github.com:neurodata/scikit-learn into obliquepr

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Add asv benchmark

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 commented:
I'm investigating the pickling issue.

The iris dataset messes up somehow. These are details of my exploration so far:

- sample numbers: `38, 85, 96, 136, 147`
- leaves: there are a total of 9 leaves and 17 nodes in the tree
- predicted leaves: they end up in the following 5 leaf indices before/after pickling:
```	
est_pred_leaves = [15 12 15 15 15]
est2_pred_leaves = [14 14 14 14 14]
```
- (not sure if this matters) the threshold in each of these leaves is “-2.0”, suggesting perhaps this is a tie-breaker issue, where the pickling is messing this up somehow
- these are the feature values for the samples that are messing up:

```
	X[diff_idx, :] == [[6.5 2.8 4.6 1.5]
	 [6.  2.7 5.1 1.6]
	 [6.3 2.3 4.4 1.3]
	 [6.2 2.2 4.5 1.5]
 	[6.3 2.5 4.9 1.5]]
```

thomasjpfan commented:
[This example](https://scikit-learn.org/stable/auto_examples/tree/plot_unveil_tree_structure.html#sphx-glr-auto-examples-tree-plot-unveil-tree-structure-py) gives an overview of how to go through the tree structure in Python.

adam2392 commented:
Crazy weird error on the first sample of the difference (i.e. sample 38 of iris): There is an incorrect addition(?) somehow...

est before pickling:
```
decision node 0 : (X_test[0, :] @ [ 1.  0. -1.  0.] = 1.9000000953674316) <= 2.6000001430511475)
decision node 1 : (X_test[0, :] @ [-1.  0.  0. -1.] = -8.0) > -8.449999809265137)
decision node 3 : (X_test[0, :] @ [ 0.  0.  0. -1.] = -1.5) > -1.6500000357627869)
decision node 9 : (X_test[0, :] @ [ 0. -1.  0.  1.] = -1.2999999523162842) > -1.3499999642372131)
decision node 11 : (X_test[0, :] @ [ 1. -1.  0.  1.] = 5.199999809265137) > 4.8999998569488525)
decision node 13 : (X_test[0, :] @ [ 1. -1.  0.  0.] = 10.199999809265137) > 9.950000286102295)
```

est after pickling:
```
decision node 0 : (X_test[0, :] @ [ 1.  0. -1.  0.] = 1.9000000953674316) <= 2.6000001430511475)
decision node 1 : (X_test[0, :] @ [-1.  0.  0. -1.] = -8.0) > -8.449999809265137)
decision node 3 : (X_test[0, :] @ [ 0.  0.  0. -1.] = -1.5) > -1.6500000357627869)
decision node 9 : (X_test[0, :] @ [ 0. -1.  0.  1.] = -1.2999999523162842) > -1.3499999642372131)
decision node 11 : (X_test[0, :] @ [ 1. -1.  0.  1.] = 5.199999809265137) > 4.8999998569488525)
decision node 13 : (X_test[0, :] @ [ 1. -1.  0.  0.] = 3.700000047683716) <= 9.950000286102295)
```

adam2392 commented:
Ah as it turns out it was a Cython issue, where I was not initializing the value of `j` to 0 when doing `for j in range(...)`. Weird that it didn't raise an error though...

adam2392 committed:
Demonstrate pickling is not an issue anymore

adam2392 committed:
Merging but pickle doesn't work still

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Fixed pickle and everything else

adam2392 committed:
Co-Authored-By: Jong Shim <jshinm@gmail.com>

moved error message and added co-author.

adam2392 committed:
Fix whatsnew

Co-Authored-By: Jong Shim <jshinm@gmail.com>

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Change c numpy to cnp

adam2392 commented:
Okay the error was a slight issue when trying to sample an oblique projection. Fixed in https://github.com/scikit-learn/scikit-learn/pull/22754/commits/a44f1b8b2140790ec7dc9f725676e68a713dbecc

adam2392 committed:
Add error message for feature importances

adam2392 committed:
Fix black

adam2392 committed:
Fix unit tests

adam2392 committed:
Adding updated tree

adam2392 committed:
Add some inline comments

adam2392 commented:
Kay this is ready for review. I've fixed the pickle issue (turns out I just needed a more robust sampling function), which has now neatly consolidated all the existing axisaligned-tree unit tests for oblique-trees. Moreover, I've added a `feature_importances` function. 

@jshinm is working with me on getting some of those examples that we discussed last time in OH: i) a notebook extending iris to demonstrate the axis/oblique split differences visually, ii) a notebook on some simulations showing oblique splits vs axis-aligned and iii) a notebook on downsampled-mnist.

@thomasjpfan would it be better to review code first, or wait for those notebooks to be "review ready"?

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Try again for azure

adam2392 committed:
Fix lint

adam2392 committed:
Apply ci fix

adam2392 committed:
Try again

adam2392 committed:
Try another rand seed

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Fix LGTM

jshinm committed:
upload iris notebook

jshinm committed:
upload simulation notebook

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Merge branch 'obliquepr' of https://github.com/neurodata/scikit-learn into obliquepr

jshinm committed:
remove notebook

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 commented:
Discussion on Discord - 6/13/22 w/ @ogrisel @glemaitre @jshinm (I don't have the other folks on GH, please feel free to add them):

1. Absolute value of the plots in addition to the delta, since maybe `max_features` -> results in diff performance in RF and OF.
2. Perform grid/random search over `max_features` and `n_estimators` to show the results of OF vs RF. Store the end results for simulation, and MNIST, and store the fitted trees on the disc. The results can be stored in a `pd.DataFrame`.

adam2392 committed:
Fix doc changes

adam2392 committed:
Fix doc changes

adam2392 committed:
Adding doc enhancements

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Fix examples

adam2392 committed:
Fix lint

adam2392 committed:
Try again

adam2392 commented:
We are working on the notebook / example that demonstrates superiority of the oblique decision trees vs axis-aligned decision trees. But based on the feedback we have received, this is ready for review! Please see the PR main post for all the information/discussion relevant for inclusion, maintenance, modularity of code, testability, and documentation.

- built docs on oblique trees: https://output.circle-artifacts.com/output/job/d244cc90-0ee6-46a0-b0f8-be0a6624ec91/artifacts/0/doc/modules/tree.html#oblique-trees (the image linked is somehow swapped, so I need to fix that, but otw the text is there).

Thanks for all the feedback and discussion to get it to this point! @glemaitre @thomasjpfan @amueller @ogrisel

adam2392 committed:
Fix doc

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Fix of

adam2392 commented:
@ogrisel I fixed the merge commit as discussed yesterday.

Here is a "brief" summary. For full details, see the PR description. Let me know if this is too long / detailed:

**What are oblique trees:** oblique trees generate splits based on combinations of features. They can sample more than `n_features` candidate splits per split node (up to 2^n in fact). Sampling random linear combinations of features implicitly captures the dominant singular vectors of the dataset with high probability, as described by the Johnson Lindenstrauss lemma. Oblique trees do not necessarily need to sample linear combinations either. They can be generalized in countless ways, which opens the door for more usage of the `sklearn.tree` module.

**When are oblique trees better than axis-aligned?** Based on the geometric motivation above, one would suspect that oblique trees fare better than axis-aligned trees when there is a high degree of noise dimensions, and/or there are optimal split hyperplanes in the data that are not aligned with the features themselves. Thus oblique trees are suspected to perform significantly better when there are a high degree of noise dimensions.

**When are axis-aligned trees better than oblique?**
Thus axis-aligned trees perform better when there are informative splits that occur along the feature axes. In general, this is a more "rare" scenario (e.g. analogously think what are the chances of sampling a 0 eigenvalue in a random matrix? measure 0). Thus, we see that oblique trees on average perform better. However, there is a small increased computational and storage cost for oblique trees: the sampling and storage of the projection vectors for each split node. Thus it is recommended to use axis-aligned trees whenever computation and storage time is a critical requirement, but always test oblique trees via cross-validation if model performance in terms of scoring metrics is preferred.

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Fixing ensemble and forest code

adam2392 committed:
Fixing black

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Fix lint

adam2392 committed:
Merge branch 'obliquepr' of https://github.com/neurodata/scikit-learn into obliquepr

adam2392 committed:
Fix lint

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Add note to doc string

adam2392 committed:
Merging

adam2392 committed:
Fix build

jjerphan commented:
Hi @adam2392, could you split this PR in several small atomic and orthogonal PRs?

There are changes (which I think we are interested in for modularity) those are orthogonal to the introduction of `ObliqueSplitter` (which we still need to discuss, IMO those better be part of another package making use of the new structure of scikit-learn tree Cython internals after have them refactored). What do you think?

adam2392 commented:
Hi @jjerphan yes I can and I have a separate proposal for BaseTree, which would also improve modularity of the Tree class. 

The plan is to:

- merge PR #24678 on modularizing criterion
- finish PR #25101 on modularizing splitter and criterion
- PR a change to modularizing trees. I can put up a sep GH issue to illustrate thoughts and a draft PR to illustrate the code proposal. 
- refactor this PR for your review (I'm still of the opinion that basic causal trees, basic oblique trees and quantile trees should be inside sklearn considering their proven fundamental importance, but happy to table discussion to the future)
- in parallel start a scikit-contrib package for the exotic tree functionalities (manifold splits, survival trees, unsupervised trees, etc.)

If there's anything I/we can do to make the reviewing process easier and faster, please let me know. We are def willing to get to the final step as fast as possible :).

adam2392 committed:
Merge branch 'main' into obliquepr

adam2392 committed:
Fixing merge

Signed-off-by: Adam Li <adam2392@gmail.com>

adam2392 committed:
Simplfy PYTHON api
